reconcile/utils/state.py

```diff
@@ -34,20 +41,44 @@ def __init__(self, integration: str, accounts: Iterable[Mapping[str, Any]],
 
         self.client = session.client('s3')
 
+        # check if the bucket exists
+        try:
+            self.client.head_bucket(Bucket=self.bucket)
+        except ClientError as details:
+            error_code = details.response.get('Error', {}).get('Code', None)
+            if error_code == '404':
+                raise StateInaccessibleException(
+                    f"bucket {self.bucket} does not exist"
+                )
+
     def exists(self, key):
         """
         Checks if a key exists in the state.
 
         :param key: key to check
 
         :type key: string
+
+        :raises StateInaccessibleException: if the bucket is missing or
+        permissions are insufficient or a general AWS error occurred
         """
         try:
             self.client.head_object(
                 Bucket=self.bucket, Key=f"{self.state_path}/{key}")
             return True
-        except ClientError:
-            return False
+        except ClientError as details:
```

Comment:
with the head_bucket check, do we really expect a 403?

Reply:
theoretically, we could see 403 when fine-grained permissions are in place (e.g. prefix conditions), but again, we really don't have to distinguish too much beyond 404. I will clean this up a bit and push another proposal